tests/foreman/cli/test_leapp_client.py

```diff
@@ -0,0 +1,259 @@
+    custom_host.install_katello_ca(satellite)
+    custom_host.register_contenthost(org=organization.label, activation_key=activation_key.name)
+    custom_host.add_rex_key(satellite=satellite)
+    assert custom_host.subscribed
+
+
+def verify_target_repo_on_satellite(
+    satellite, content_view, organization, lifecycle_env, target_rhel
+):
+    """Verify target rhel version repositories has enabled on Satellite Server"""
+    cmd_out = satellite.execute(
+        f"hammer repository list --search 'content_label ~ {target_rhel}' "
+        f"--content-view {content_view.name} --organization '{organization.name}' "
+        f"--lifecycle-environment '{lifecycle_env.name}'"
+    )
+    assert cmd_out.status == 0
+    if 'rhel-9' in target_rhel:
+        assert ("AppStream RPMs 9" in cmd_out.stdout) and ("BaseOS RPMs 9" in cmd_out.stdout)
+    else:
+        # placeholder for target_rhel - rhel-8
+        pass
+
+
+def precondition_check_upgrade_and_install_leapp_tool(custom_host, source_rhel):
+    """Clean-up directory, set rhel release version, update system and install leapp tool"""
+    # Remove directory if in-place upgrade already performed from RHEL7 to RHEL8
+    custom_host.run("rm -rf /root/tmp_leapp_py3")
+    custom_host.run("dnf clean all")
+    custom_host.run("dnf repolist")
+    custom_host.run(f"subscription-manager release --set {source_rhel}")
+    assert custom_host.run("dnf update -y").status == 0
+    assert custom_host.run("dnf install leapp-upgrade -y").status == 0
+
+
+def fix_inhibitors(custom_host, source_rhel):
+    """Fixing inhibitors to avoid hard stop of Leapp tool execution"""
+    if '8' in source_rhel:
+        # 1. Firewalld Configuration AllowZoneDrifting Is Unsupported
+        custom_host.run(
+            'sed -i "s/^AllowZoneDrifting=.*/AllowZoneDrifting=no/" /etc/firewalld/firewalld.conf'
+        )
+        # 2. Newest installed kernel not in use
+        if custom_host.run('needs-restarting -r').status == 1:
+            custom_host.power_control(state='reboot', ensure=True)
+    else:
+        # placeholder for source_rhel - 7
+        pass
+
+
+@pytest.fixture(scope="module")
+def setup_env(module_target_sat, module_sca_manifest_org):
+    """Creating essential things and returning in form of dictiory for use"""
+    # 1. Import a subscription manifest
+    lc_env = module_target_sat.api.LifecycleEnvironment(
+        organization=module_sca_manifest_org
+    ).create()
+    c_view = module_target_sat.api.ContentView(organization=module_sca_manifest_org).create()
+    return {
+        'organization': module_sca_manifest_org,
+        'satellite': module_target_sat,
+        'environment': lc_env,
+        'contentview': c_view,
+    }
+
+
+@pytest.mark.no_containers
+@pytest.mark.parametrize(
+    'custom_host',
+    [
+        {'deploy_rhel_version': '8.8'},
+    ],
+    ids=['RHEL8.8'],
+    indirect=True,
+)
+@pytest.mark.parametrize('target_rhel', ['rhel-9'])
```

Comment:
As you'll be upgrading it rhel9, and skip level upgrade isn't possible, so why parametrizing it? :)

Reply:
- The earlier plan was to make the same test case parameterize so both upgrade paths would work, later on, realizes that it would be confusing & decided to make separate test.
- Also `target_rhel` is indicating major version not a actual OS version, it doesn't make any sense to use here in test
- So I will remove this marker and use plane text string wherever requires.